File: src/prbm_stackqueue/SteelPipe.java
package prbm_stackqueue;

public class SteelPipe {

	public static void main(String[] args) {
		String arrangement = "()(((()())(())()))(())";
		System.out.println(solution(arrangement));
		
		
	}
	
    public static int solution(String arrangement) {
    	int length = arrangement.length();
    	int stack = 0;    
    	int cnt = 0;
    	boolean isOpen = true;
    	
    	for(int i=0; i<length; i++) {
    		if (arrangement.charAt(i) == '(' ) {
    			++stack;
    			isOpen = true;
    		} else {
    			--stack;
    			if(isOpen == true) {
    				cnt+=stack;
    			} else {
    				++cnt;
    			}    			
    			isOpen = false;
    		}
    	}
    	
        return cnt;
    }
}

File: src/lan_python/prbm_stackqueue/SteelPipe.py
def solution(arrangement):
    stack = 0
    answer = 0
    last = True # True == '(' , False = ')'
    for idx, val in enumerate(arrangement) :
        
        if val == '(' :
            last = True
            stack += 1
        else : # val == ')'
            ## answer에 더해주는 행위는 )가 나올 때만 해주는게 포인트
            if last == True :
                answer += stack-1 # stack-1값만큼 더해주어야 한다.
            else :
                answer += 1
            last = False
            stack -= 1

    return answer